pkg/network/common/dns.go

```diff
@@ -122,58 +133,110 @@ func (d *DNS) updateOne(dns string) (bool, error) {
 	return changed, nil
 }
 
-func (d *DNS) getIPsAndMinTTL(domain string) ([]net.IP, time.Duration, error) {
+func (d *DNS) doOneQuery(server, domain string, rtype uint16) ([]net.IP, int, error) {
 	ips := []net.IP{}
-	ttlSet := false
-	var ttlSeconds uint32
+	ttl := defaultTTL
 
-	for _, server := range d.nameservers {
-		msg := new(dns.Msg)
-		msg.SetQuestion(dns.Fqdn(domain), dns.TypeA)
+	msg := new(dns.Msg)
+	msg.SetQuestion(dns.Fqdn(domain), rtype)
 
-		dialServer := server
-		if _, _, err := net.SplitHostPort(server); err != nil {
-			dialServer = net.JoinHostPort(server, d.port)
-		}
-		c := new(dns.Client)
-		c.Timeout = 5 * time.Second
-		in, _, err := c.Exchange(msg, dialServer)
-		if err != nil {
-			return nil, defaultTTL, err
-		}
-		if in != nil && in.Rcode != dns.RcodeSuccess {
-			return nil, defaultTTL, fmt.Errorf("failed to get a valid answer: %v", in)
+	c := new(dns.Client)
+	c.Timeout = d.timeout
+	in, _, err := c.Exchange(msg, server)
+	if in == nil || err != nil {
+		return ips, ttl, err
+	}
+	if in.Rcode != dns.RcodeSuccess {
+		return ips, ttl, fmt.Errorf("failed to get a valid answer: %v", in)
+	}
+
+	for _, a := range in.Answer {
+		aTTL := int(a.Header().Ttl)
+		if aTTL < ttl && aTTL != 0 {
+			ttl = aTTL
 		}
 
-		if in != nil && len(in.Answer) > 0 {
-			for _, a := range in.Answer {
-				if !ttlSet || a.Header().Ttl < ttlSeconds {
-					ttlSeconds = a.Header().Ttl
-					ttlSet = true
-				}
-
-				switch t := a.(type) {
-				case *dns.A:
-					ips = append(ips, t.A)
-				}
+		switch t := a.(type) {
+		case *dns.A:
+			if rtype == dns.TypeA {
+				ips = append(ips, t.A)
+			}
+		case *dns.AAAA:
+			if rtype == dns.TypeAAAA {
+				ips = append(ips, t.AAAA)
 			}
 		}
 	}
 
-	if !ttlSet || (len(ips) == 0) {
-		return nil, defaultTTL, fmt.Errorf("IPv4 addr not found for domain: %q, nameservers: %v", domain, d.nameservers)
-	}
+	return ips, ttl, nil
+}
 
-	ttl, err := time.ParseDuration(fmt.Sprintf("%ds", ttlSeconds))
-	if err != nil {
-		utilruntime.HandleError(fmt.Errorf("Invalid TTL value for domain: %q, err: %v, defaulting ttl=%s", domain, err, defaultTTL.String()))
-		ttl = defaultTTL
+func (d *DNS) queryServer(nameserver, domain string) ([]net.IP, int, error) {
+	if d.ipv4 && !d.ipv6 {
+		// Single-stack IPv4
+		return d.doOneQuery(nameserver, domain, dns.TypeA)
+	} else if d.ipv6 && !d.ipv4 {
+		// Single-stack IPv6
+		return d.doOneQuery(nameserver, domain, dns.TypeAAAA)
 	}
-	if ttl == 0 {
-		ttl = defaultTTL
+	// else dual stack
+	ips := []net.IP{}
+	ttl := defaultTTL
+	errs := make(chan error)
+	var mutex sync.Mutex
+
+	go func() {
```

Comment:
I'm a bit concerned about this blocking...
If one of the petition needs to wait 5 seconds for a retry then the other will be blocked regardless which means instead of having just one family outdated for 5 seconds you have both.

That's a pretty big change and I don't think it has a huge but could we consider add a fixme here?

Reply:
Well... the really big problem would be if the DNS server is old and broken and ignores AAAA requests, in which case we'll time out every time. But for the use case here (egress firewall) we _have_ to wait for both answers so I'm not sure there's anything better we could do...

Reply:
> But for the use case here (egress firewall) we have to wait for both answers so I'm not sure there's anything better we could do...

Small delays when updating the flows timely can cause a significant amount of drops: https://bugzilla.redhat.com/show_bug.cgi?id=1850060 and more often than not, A and AAAA records have different TTLs.

In other words, if we're doing the query because we need to update the A record and the AAAA query delays for whatever reason, we'll block the updates for the A record while we wait for an AAAA record which hasn't changed.

However, this involves modifying EgressDNS.Sync() and EgressDNS.GetNextQueryTime() (which already are getting rewritten for IPv4 alone https://github.com/openshift/sdn/pull/168) and that's a pretty big change for an issue that for all we know is only potential.